docs/proposals/provider-agnostic-dns-health-checks.md

```diff
@@ -0,0 +1,140 @@
+a health check is created based on the health check configuration in the DNSPolicy.
+
+A `failureThreshold` field will be added to the health spec, allowing users to configure a number of consecutive health 
+check failures that must be observed before the endpoint is considered unhealthy.
+
+Example DNS Policy with a defined health check.
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSPolicy
+metadata:
+  name: prod-web
+  namespace: multi-cluster-gateways
+spec:
+  healthCheck:
+    endpoint: /health
+    failureThreshold: 5
+    port: 443
+    protocol: https
+    additionalHeaders:
+      - name: "..."
+        value: "..."
+      - name: "..."
+        value: "..."
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: Gateway
+    name: prod-web
+    namespace: multi-cluster-gateways
+```
+#### `DNSHealthCheckProbe` resource
+
+The DNSHealthCheckProbe resource configures a health probe in the controller to perform the health checks against an 
+identified final A or CNAME endpoint. When created by the controller as a result of a DNS Policy, this will have an 
+owner ref of the DNS Policy that caused it to be created.
+
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSHealthCheckProbe
+metadata:
+  name: example-probe
+spec:
+  port: "..."
+  host: “...”
+  address: "..."
+  path: "..."
+  protocol: "..."
+  interval: "..."
+  additionalHeaders:
+    - name: "..."
+      value: "..."
+    - name: "..."
+      value: "..."
+status:
+  healthy: true
+  consecutiveFailures: 0
+  reason: ""
+  lastCheck: "..."
+```
+
+#### Spec Fields Definition
+- **Port** The port to use
+- **Address** The address to connect to (e.g. IP address or hostname of a clusters loadbalancer)
+- **Host** The host to request in the Host header
+- **Path** The path to request
+- **Protocol** The protocol to use for this request
+- **Interval** How frequently this check would ideally be executed.
+- **AdditionalHeaders** Optional headers and their values that can be specified to ensure the health check is successful.
+
+The reconciliation of this resource results in the configuration of a health probe, which targets the endpoint and 
+updates the status. The status is propagated to the providerSpecific status of the equivalent endpoint in the DNSRecord
+
+### Changes to current controllers
+
+In order to support this new feature, the following changes in the behaviour of the controllers are proposed.
+
+#### DNSPolicy controller
+
+Currently, the reconciliation loop of this controller creates health checks in the configured DNS provider 
+(Route53 currently) based on the spec of the DNSPolicy, separately from the reconciliation of the DNSRecords. 
+The proposed change is to reconcile health check probe CRs based on the combination of DNS Records and DNS Policies.
+
+Instead of Route53 health checks, the controller will create `DNSHealthCheckProbe` resources.
```

Comment:
Is it expected a user (gateway admin or application developer) would create a DNSHealthCheckProbe?

Reply:
no, these are created by the controller as a result of the settings in a dns policy.